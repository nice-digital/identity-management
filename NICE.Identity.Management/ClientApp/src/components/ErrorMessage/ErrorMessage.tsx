import React from "react";

import { Alert } from "@nice-digital/nds-alert";

type ErrorMessageProps = {
	error: Error;
	children?: React.ReactNode;
};

export const ErrorMessage = (props: ErrorMessageProps): React.ReactElement => (
	<Alert type="error">
		{props.children ? (
			props.children
		) : (
			<>
				<h2>Whoops</h2>
				<p>Sorry, there&apos;s been an error!</p>
			</>
		)}

		{process.env.NODE_ENV === "development" && (
			<dl>
				<dt>Message:</dt>
				<dd>{props.error.message}</dd>
				<dt>Stack:</dt>
				<dd>{props.error.stack}</dd>
			</dl>
		)}
	</Alert>
);
