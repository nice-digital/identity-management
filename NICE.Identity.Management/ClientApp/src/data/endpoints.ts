const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
