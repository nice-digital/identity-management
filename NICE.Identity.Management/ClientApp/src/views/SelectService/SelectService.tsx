import React, { Component } from "react";
import { RouteComponentProps, Link } from "react-router-dom";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { PageHeader } from "@nice-digital/nds-page-header";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { StackedNav, StackedNavLink } from "@nice-digital/nds-stacked-nav";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";

import { Endpoints } from "../../data/endpoints";
import { fetchData } from "../../helpers/fetchData";
import { isDataError } from "../../helpers/isDataError";
import { ServiceType, UserType } from "../../models/types";

type TParams = { id: string };

type SelectServiceProps = Record<string, unknown> &
	RouteComponentProps<TParams>;

type SelectServiceState = {
	user: UserType;
	services: Array<ServiceType>;
	error?: Error;
	isLoading: boolean;
};

export class SelectService extends Component<
	SelectServiceProps,
	SelectServiceState
> {
	constructor(props: SelectServiceProps) {
		super(props);
		this.state = {
			user: {} as UserType,
			services: [],
			isLoading: true,
		};
		document.title = "NICE Accounts - Select service";
	}

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const user = await fetchData(Endpoints.user(this.props.match.params.id));
		const services = await fetchData(Endpoints.servicesList);

		if (isDataError(user)) {
			this.setState({ error: user });
		}
		if (isDataError(services)) {
			this.setState({ error: services });
		}

		this.setState({ user, services, isLoading: false });
	}

	render(): JSX.Element {
		const { user, services, error, isLoading } = this.state;
		const { id } = this.props.match.params;

		let nameBreadcrumb = `${user.firstName} ${user.lastName}`;

		if (isLoading) {
			nameBreadcrumb = "Loading user details";
		}
		if (error) {
			nameBreadcrumb = "Error";
		}

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb to="/users" elementType={Link}>
						Users
					</Breadcrumb>
					<Breadcrumb to={`/users/${id}`} elementType={Link}>
						{nameBreadcrumb}
					</Breadcrumb>
					<Breadcrumb>Select service</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader preheading="Select service" heading="Service list" />

						<Grid>
							<GridItem cols={12}>
								{isLoading ? (
									<p>Loading...</p>
								) : (
									<StackedNav
										data-qa-sel="service-stacked-nav"
										aria-label="Services"
									>
										{services.map((service) => {
											return (
												<StackedNavLink
													data-qa-sel="service-stacked-nav-link"
													destination={`${this.props.match.url}/${service.id}/environments`}
													elementType={Link}
													key={service.id}
												>
													{service.name}
												</StackedNavLink>
											);
										})}
									</StackedNav>
								)}
							</GridItem>
						</Grid>
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
