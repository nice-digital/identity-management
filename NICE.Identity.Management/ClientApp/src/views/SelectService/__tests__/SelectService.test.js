import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";

import { nextTick } from "../../../utils/nextTick";
import singleUser from "./singleUser.json";
import services from "./services.json";
import { SelectService } from "../SelectService";

import { Endpoints } from "../../../data/endpoints";
import { ErrorMessage } from "../../../components/ErrorMessage/ErrorMessage";

describe("SelectService", () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});

	it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(services));
		const wrapper = shallow(<SelectService match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetchData during componentDidMount", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(services));
		const wrapper = mount(<MemoryRouter><SelectService match={match} /></MemoryRouter>);
		const spy = jest.spyOn(wrapper.instance(), "componentDidMount");
		wrapper.instance().componentDidMount();
		await nextTick();
		wrapper.update();
		expect(spy).toHaveBeenCalled();
		expect(fetch.mock.calls.length).toEqual(2);
		expect(fetch.mock.calls[0][0]).toEqual(Endpoints.user(match.params.id));
		expect(fetch.mock.calls[1][0]).toEqual(Endpoints.servicesList);
		spy.mockClear();
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(services));
		const wrapper = shallow(<SelectService match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when user fetchData function returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		fetch.mockResponseOnce(JSON.stringify(services));
		const wrapper = mount(
			<MemoryRouter>
				<SelectService match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when user fetchData function returns 401 error", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		fetch.mockResponseOnce(JSON.stringify(services));
		const wrapper = mount(
			<MemoryRouter>
				<SelectService match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		const wrapper = mount(
			<MemoryRouter>
				<SelectService match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 401 error", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		const wrapper = mount(
			<MemoryRouter>
				<SelectService match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});
});
