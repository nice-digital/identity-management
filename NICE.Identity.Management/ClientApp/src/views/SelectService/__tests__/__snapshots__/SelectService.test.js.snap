// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot after data has been loaded 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Kristin Patrick
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select service
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Kristin Patrick"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Select service"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      <span
        class="page-header__pre-heading"
      >
        Select service
      </span>
      Service list
    </h1>
  </div>
  <div
    class="grid"
  >
    <div
      data-g="12"
    >
      <nav
        aria-label="Services"
        class="stacked-nav"
        data-qa-sel="service-stacked-nav"
      >
        <ul
          class="stacked-nav__list"
        >
          <li
            class="stacked-nav__list-item"
            data-qa-sel="service-stacked-nav-link"
          >
            <a
              aria-current="false"
              href="/1/environments"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                EPPI Reviewer
              </span>
            </a>
          </li>
          <li
            class="stacked-nav__list-item"
            data-qa-sel="service-stacked-nav-link"
          >
            <a
              aria-current="false"
              href="/2/environments"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                Comment Collection
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;

exports[`should show error message when services fetchData function returns 401 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Error
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select service
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Select service"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when services fetchData function returns 500 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Error
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select service
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Select service"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when user fetchData function returns 401 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Error
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select service
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Select service"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when user fetchData function returns 500 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Error
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select service
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Select service"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;
